DataGrid: Change default values of TemplateColumn

dennisrahmen opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
The `TemplateColumn` is generally used to add "Action" columns like editing, select and hierarchy.
Currently, the defaults are just the same as for the normal column, which is suboptimal.
In the most common use cases most features are not used or need to be setup first anyway, like with `SortBy`.

Some of these features like drag and drop add spacing to the column, which is particularly noticeable for select and hierarchy columns.

Furthermore, some features could not been turned off per column, like drag and drop for the hierarchy and select columns.

I also changed the default for 'ShowInFooter' for the SelectColumn since I have seen multiple people complain about the hole row just for the one checkbox.
I would remove that completely to be honest, no other columns support this, and you could always make the header sticky with 'FixedHeader' and have a better experience.

Examples of space savings:
![image](https://github.com/MudBlazor/MudBlazor/assets/61705878/779a0805-8b7a-4435-924e-85a8336564e5)

![image](https://github.com/MudBlazor/MudBlazor/assets/61705878/a6081c71-5c34-4b7f-b650-447aa74486cd)

![image](https://github.com/MudBlazor/MudBlazor/assets/61705878/e5d962fe-860b-45ee-89ae-c7264f99e4df)

![image](https://github.com/MudBlazor/MudBlazor/assets/61705878/4660fa0e-8b3d-405d-a58c-924eb117fa64)


## How Has This Been Tested?
visually

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

dennisrahmen committed:
Change default values of TemplateColumn

dennisrahmen commented:
If this PR gets merged the following default values have changed to `false` and need to be added to the migration guide:

### Template Column

- ShowColumnOptions
- Resizable
- DragAndDropEnabled
- Sortable
- Filterable

### Select Column

- DragAndDropEnabled (change inherited from Template column)
- ShowInFooter

### Hierarchy Column (changes inherited from Template column)

- DragAndDropEnabled

dennisrahmen commented:
@tjscience last year we talked about this change here: https://github.com/MudBlazor/MudBlazor/pull/6049#issuecomment-1452741566
I hope we had the same thing in mind by setting sensible defaults.

dennisrahmen committed:
Fix formatting

dennisrahmen committed:
Fix tests

dennisrahmen commented:
@henon @danielchalmers @ScarletKuro ready for review ✌️

dennisrahmen committed:
Add missing DragAndDropEnabled in component

ScarletKuro commented:
I will let @tjscience to decide, I'm unsure about the `ShowInFooter="false"` default as it leaves awkward space in footer, also it's been default for a very long time. I only pretty much agree with `Fiterable` as `TemplateColumn` is not attached to any data and filtering is not possible as result confusing people

tjscience reviewed (approved):
I think everything looks good here. Thanks, @dennisrahmen!

dennisrahmen commented:
@ScarletKuro the "awkward space in footer" was there already and is not even dependent on the `ShowInFooter` setting, as it is introduced though the spacing of the checkboxes.

I wanted to open a separate pr for that.

ScarletKuro commented:
@dennisrahmen can you write down all breaking change of that PR for the https://github.com/MudBlazor/MudBlazor/issues/8447
in similar format so we could just copy paste it? Thanks.

dennisrahmen commented:
@ScarletKuro @henon I took the hole block so you can just copy paste.

```
## MudDataGrid

- Removed `FilterDefinition.GenerateFilterExpression()` replace with `FilterExpressionGenerator.GenerateExpression`
- Removed other old APIs from `FilterDefinition` that was replaced by `IFilterDefinition` API.
- `FilterContext.FilterActions` all properties are marked as required.
- `FooterContext.FooterActions` all properties are marked as required.
- `HeaderContext.SetSelectAllAsync` all properties are marked as required.
- `HeaderContext.IsAllSelected` changed type from `bool` to `bool?`.
- **Column**: replace `IsEditable` with `Editable`
- **TemplateColumn**
  - `ShowColumnOptions` changed default to false
  - `Resizable` changed default to false
  - `DragAndDropEnabled` changed default to false
  - `Sortable` changed default to false
  - `Filterable` changed default to false
- **SelectColumn**
  - `DragAndDropEnabled` changed default to false
  - `ShowInFooter` changed default to false
- **HierarchyColumn**
  - `DragAndDropEnabled` changed default to false

For more details see https://github.com/MudBlazor/MudBlazor/pull/8548, https://github.com/MudBlazor/MudBlazor/pull/8317, https://github.com/MudBlazor/MudBlazor/pull/8892, https://github.com/MudBlazor/MudBlazor/pull/9142
```

henon commented:
Thanks Dennis
